File: July-LeetCoding-Challenge/27-Construct-Binary-Tree-from-Inorder-and-Postorder-Traversal/Construct-Binary-Tree-from-Inorder-and-Postorder-Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public TreeNode BuildTree(int[] inorder, int[] postorder) {
        if(inorder == null || postorder == null || inorder.Length != postorder.Length) return null;
        var dict = new Dictionary<int, int>();
        for(int i=0; i<inorder.Length; i++){
            dict.Add(inorder[i], i);
        }
        
        return BuildTree(inorder, 0, inorder.Length, postorder, 0, postorder.Length-1, dict);
    }
    
    private TreeNode BuildTree(int[] inorder, int ib, int ie, int[] postorder, int pb, int pe, Dictionary<int, int> dict)
    {
        if(pb>pe || ib>ie) return null;
        var root = new TreeNode(postorder[pe]);
        var rootIdx = dict[postorder[pe]];
        var left = BuildTree(inorder, ib, rootIdx-1, postorder, pb, pb+rootIdx-ib-1, dict);
        var right = BuildTree(inorder, rootIdx+1, ie, postorder, pb+rootIdx-ib, pe-1, dict);
        root.left = left;
        root.right = right;
        return root;
    }
}

File: July-LeetCoding-Challenge/27-Construct-Binary-Tree-from-Inorder-and-Postorder-Traversal/Construct-Binary-Tree-from-Inorder-and-Postorder-Traversal.kt
package binary_tree

class ConstructBinaryTreefromInorderandPostorderTraversalKotlin106 {
    fun buildTree(inorder: IntArray, postorder: IntArray): TreeNode? =
        construct(
            inorder,
            0,
            inorder.size - 1,
            postorder,
            0,
            postorder.size - 1
        )

    private fun construct(
        inorder: IntArray,
        iStart: Int,
        iEnd: Int,
        postorder: IntArray,
        pStart: Int,
        pEnd: Int
    ): TreeNode? {
        if (iStart > iEnd || pStart > pEnd) {
            return null
        }
        val treeNode = TreeNode(postorder[pEnd])
        var index = iStart
        while (iStart < inorder.size && inorder[index] != treeNode.`val`) {
            ++index
        }
        treeNode.left = construct(inorder, iStart, index - 1, postorder, pStart, pStart + index - 1 - iStart)
        treeNode.right = construct(inorder, index + 1, iEnd, postorder, pStart + index - iStart, pEnd - 1)
        return treeNode
    }

    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }
}

fun main() {
    val solution = ConstructBinaryTreefromInorderandPostorderTraversalKotlin106()
    val result = solution.buildTree(
        intArrayOf(9, 3, 15, 20, 7),
        intArrayOf(9, 15, 7, 20, 3)
    )
    println(result)
}